src/rules/attribute-names.ts

```diff
@@ -0,0 +1,69 @@
+/**
+ * @fileoverview Enforces attribute naming conventions
+ * @author James Garbutt <https://github.com/43081j>
+ */
+
+import {Rule} from 'eslint';
+import * as ESTree from 'estree';
+import {getPropertyMap, isLitClass} from '../util';
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+const rule: Rule.RuleModule = {
+  meta: {
+    docs: {
+      description: 'Enforces attribute naming conventions',
+      recommended: true,
+      url: 'https://github.com/43081j/eslint-plugin-lit/blob/master/docs/rules/attribute-names.md'
+    },
+    schema: [],
+    messages: {
+      casedAttribute:
+        'Attributes are case-insensitive and therefore should be ' +
+        'defined in lower case',
+      casedPropertyWithoutAttribute:
+        'Property has non-lowercase casing but no attribute. It should ' +
+        'instead have an explicit `attribute` set to the lower case ' +
+        'name (usually snake-case)'
+    }
+  },
+
+  create(context): Rule.RuleListener {
+    return {
+      ClassDeclaration: (node: ESTree.Class): void => {
+        if (isLitClass(node)) {
+          const propertyMap = getPropertyMap(node);
+
+          for (const [prop, propConfig] of propertyMap.entries()) {
+            if (!propConfig.attribute) {
+              continue;
+            }
+
+            if (!propConfig.attributeName) {
+              if (prop.toLowerCase() !== prop) {
+                context.report({
+                  node: propConfig.key,
+                  messageId: 'casedPropertyWithoutAttribute'
+                });
+              }
```

Comment:
I'm just wondering how necessary this part is? Whilst this explicitly mandates it, from my understanding, lit already implicitly lowercases property names already.

> By default, Lit creates a corresponding observed attribute for all public reactive properties. The name of the observed attribute is the property name, lowercased:
https://lit.dev/docs/components/properties/

Is it therefore adding a rules where one is not needed?

Reply:
attributes are always lowercase, so even if lit does nothing a `camelCase` property will be available as `camelcase`

this was more to push people to snake-case it (which the error message suggests). since i've quite often seen something like `someLongCamelCaseProp` used in HTML (`somelongcamelcasedprop`), which would be more readable as `some-long-camel-cased-prop`. the thought here was, by pushing people to explicitly define the attribute, they'll probably realise at the same time they should make it snake-case (or at least put some thought into what'd be a readable attribute, rather than no thought)

Reply:
Reasoning is sound - works for me

Reply:
Reasoning is sound - works for me